{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "196ef039-a6bb-4b95-8792-20d7c478b4aa",
   "metadata": {},
   "source": [
    "# 1. Reads Assignment - HeLa FUCCI Cells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "543da5e2",
   "metadata": {},
   "source": [
    "2023-03-23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29f68cd6-7e52-4e51-9781-08bee68d1807",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load libraries\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "from natsort import natsorted\n",
    "from scipy.io import loadmat, savemat\n",
    "from skimage.filters import threshold_otsu\n",
    "from skimage.color import label2rgb\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "# Load starmap python library \n",
    "from starmap.sequencing import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff4949f6-b3d2-4831-a0ef-3ad5343b8d38",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e40368a2-1816-4b99-b8c9-c45ed2891688",
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import wraps\n",
    "from time import time\n",
    "\n",
    "# Timer\n",
    "def timer(func):\n",
    "    @wraps(func)\n",
    "    def _time_it(*args, **kwargs):\n",
    "        start = int(round(time() * 1000))\n",
    "        try:\n",
    "            return func(*args, **kwargs)\n",
    "        finally:\n",
    "            end_ = int(round(time() * 1000)) - start\n",
    "            end_ = round(end_ / 1000, 4)\n",
    "            print(f\"Total execution time: {end_ if end_ > 0 else 0} s\")\n",
    "    return _time_it\n",
    "\n",
    "\n",
    "@timer\n",
    "# Load reads and their positions from mat file (generated by matlab preprocessing pipeline)\n",
    "def load_reads(fpath, reads_file):\n",
    "    S = loadmat(os.path.join(fpath, reads_file))\n",
    "    bases = [str(i[0][0]) for i in S[\"merged_reads\"]]\n",
    "    points = S[\"merged_points\"]\n",
    "    temp = np.zeros(points.shape)\n",
    "    temp[:, 0] = np.round(points[:, 1]-1)\n",
    "    temp[:, 1] = np.round(points[:, 0]-1)\n",
    "    temp[:, 2] = np.round(points[:, 2]-1)\n",
    "    \n",
    "    print(f\"Number of reads: {len(bases)}\")\n",
    "    \n",
    "    return bases, temp\n",
    "\n",
    "\n",
    "# Load gene barcode table from genes.csv\n",
    "def load_genes(fpath):\n",
    "    genes2seq = {}\n",
    "    seq2genes = {}\n",
    "    with open(os.path.join(fpath, \"genes.csv\"), encoding='utf-8-sig') as f:\n",
    "        for l in f:\n",
    "            fields = l.rstrip().split(\",\")\n",
    "            genes2seq[fields[0]] = \"\".join([str(s+1) for s in encode_SOLID(fields[1][::-1])])\n",
    "            seq2genes[genes2seq[fields[0]]] = fields[0]\n",
    "    return genes2seq, seq2genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba234b87-d507-4b8e-99f9-c49a02efe3a7",
   "metadata": {},
   "source": [
    "## Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e3c617d-9031-49e9-9f20-176697797945",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IO path \n",
    "input_path = 'path_to_preprocessed_dataset'\n",
    "\n",
    "out_path = os.path.join(input_path, 'output')\n",
    "if not os.path.exists(out_path):\n",
    "    os.mkdir(out_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "510e9e6a-3ed0-491f-b896-e1c6cf2e8989",
   "metadata": {},
   "source": [
    "## Run pipeline for individual sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "748d3415-2f0c-41d5-9243-3bfb6fd18881",
   "metadata": {},
   "outputs": [],
   "source": [
    "# parameters \n",
    "current_sample = 'ribomap'\n",
    "# current_sample = 'starmap'\n",
    "\n",
    "print(f\"Current sample: {current_sample}\")\n",
    "\n",
    "# Load genes\n",
    "genes2seqs, seqs2genes = load_genes(input_path)\n",
    "\n",
    "# Load reads \n",
    "bases, points = load_reads(os.path.join(input_path, 'merged_spots', current_sample), \"merged_goodPoints_max3d.mat\")\n",
    "bases = np.array([seqs2genes[x] for x in bases])\n",
    "points = np.array(points)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12b42abe-635f-4a1b-b93b-874e8646e984",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run reads assignment for each cellular compartment \n",
    "structure_dict = {'whole_cell': 'cell.tif',\n",
    "                 'nucleus': 'nuclei.tif',\n",
    "                 'cytoplasm': 'cyto.tif',\n",
    "                 'outer_cytoplasm': 'outer_cyto.tif',\n",
    "                 'er': 'er.tif'}\n",
    "\n",
    "total_cells = None\n",
    "\n",
    "expr_out_path = os.path.join(out_path, current_sample)\n",
    "if not os.path.exists(expr_out_path):\n",
    "    os.mkdir(expr_out_path)\n",
    "        \n",
    "expr_dict = {}\n",
    "for current_structure in structure_dict.keys():\n",
    "    print(f\"====Processing: {current_structure}====\")\n",
    "    # Path\n",
    "    seg_path = os.path.join(input_path, 'segmentation', current_sample)\n",
    "    img_path = os.path.join(input_path, 'images', current_sample)\n",
    "\n",
    "    # Load segmentation\n",
    "    current_seg = load_label_image(seg_path, fname=structure_dict[current_structure])\n",
    "\n",
    "    # # QC plots on whole cell segmentation \n",
    "    # if current_structure == 'whole_cell':\n",
    "        \n",
    "    #     # Load image\n",
    "    #     current_img = load_label_image(seg_path, fname='overlay_label.tiff')\n",
    "\n",
    "    #     # Set figsize\n",
    "    #     figsize = (np.floor(current_img.shape[1] / 1000 * 5), np.floor(current_img.shape[0] / 1000 * 5))\n",
    "\n",
    "    #     # Plot cell number \n",
    "    #     t_size = 10\n",
    "    #     plt.figure(figsize=figsize)\n",
    "    #     plt.imshow(current_img > 0, cmap='Greys')\n",
    "    #     for i, region in enumerate(regionprops(current_seg)):\n",
    "    #         plt.plot(region.centroid[1], region.centroid[0], '.', color='red', markersize=4)\n",
    "    #         plt.text(region.centroid[1], region.centroid[0], str(region.label - 1), fontsize=t_size, color='red')\n",
    "    #     plt.axis('off')\n",
    "    #     plt.savefig(os.path.join(seg_path, f\"cell_nums_{current_structure}.png\"))\n",
    "    #     plt.clf()\n",
    "    #     plt.close()\n",
    "    \n",
    "    #     # Plot dots on segmentation mask\n",
    "    #     plt.figure(figsize=figsize)\n",
    "    #     plt.imshow(current_img > 0, cmap='gray')\n",
    "    #     plt.plot(points[:, 1], points[:, 0], '.', color='red', markersize=1)\n",
    "    #     plt.axis('off')\n",
    "    #     print(f\"Saving points_seg.png\")\n",
    "    #     plt.savefig(os.path.join(seg_path, f\"points_seg_{current_structure}.png\"))\n",
    "    #     plt.clf()\n",
    "    #     plt.close()\n",
    "    \n",
    "    # Reads assignment to current_structure\n",
    "    current_out_path = os.path.join(expr_out_path, current_structure)\n",
    "    if not os.path.exists(current_out_path):\n",
    "        os.mkdir(current_out_path)\n",
    "\n",
    "    points = points.astype(int)\n",
    "    reads_assignment = current_seg[points[:, 2], points[:, 0], points[:, 1]]\n",
    "\n",
    "    if not total_cells:\n",
    "        total_cells = len(np.unique(current_seg)) - 1\n",
    "        print(f\"Total number of cells: {total_cells}\")\n",
    "\n",
    "    genes = genes2seqs.keys()\n",
    "    cell_by_barcode = np.zeros((total_cells, len(genes)))\n",
    "    genes_to_index = {}  # map from sequence to index into matrix\n",
    "\n",
    "    for i, k in enumerate(genes):\n",
    "        genes_to_index[k] = i\n",
    "        \n",
    "    # Get region information \n",
    "    print('Getting region information...')\n",
    "    region_labels = []\n",
    "    areas_dict = []\n",
    "    cell_locs_dict = []\n",
    "    for i, region in enumerate(tqdm(regionprops(current_seg))):\n",
    "        region_labels.append(region.label)\n",
    "        areas_dict.append(region.area)\n",
    "        cell_locs_dict.append(region.centroid)\n",
    "        \n",
    "        # areas_dict[region.label] = region.area\n",
    "        # cell_locs_dict[region.label] = region.centroid\n",
    "    \n",
    "    # Iterate through cells\n",
    "    print('Iterate cells...')\n",
    "    areas = []\n",
    "    cell_locs = []\n",
    "    for i in tqdm(range(total_cells)):\n",
    "        current_label = i+1\n",
    "        if current_label in region_labels:\n",
    "            areas.append(areas_dict[region_labels.index(current_label)])\n",
    "            cell_locs.append(cell_locs_dict[region_labels.index(current_label)])\n",
    "        else:\n",
    "            areas.append(0)\n",
    "            cell_locs.append([0, 0, 0])\n",
    "            \n",
    "        assigned_reads = bases[np.argwhere(reads_assignment == current_label).flatten()]\n",
    "        for j in assigned_reads:\n",
    "            if j in genes_to_index:\n",
    "                cell_by_barcode[i, genes_to_index[j]] += 1\n",
    "\n",
    "    cell_locs = np.array(cell_locs).astype(int)\n",
    "    curr_meta = pd.DataFrame({'sample': 'STARmap', 'area': areas,\n",
    "                              'x':cell_locs[:, 1], 'y':cell_locs[:, 2], 'z':cell_locs[:, 0]})\n",
    "\n",
    "    # Output\n",
    "    with open(os.path.join(current_out_path, \"log.txt\"), 'w') as f:\n",
    "        msg = \"{:.2%} percent [{} out of {}] reads were assigned to {} cells\".format(cell_by_barcode.sum()/len(bases), cell_by_barcode.sum(), len(bases), total_cells)\n",
    "        print(msg)\n",
    "        f.write(msg)\n",
    "        \n",
    "    np.savetxt(os.path.join(current_out_path, \"cell_barcode_count.csv\"), cell_by_barcode.astype(np.int), delimiter=',', fmt=\"%d\")\n",
    "    cell_barcode_names = pd.DataFrame({'seq': list(seqs2genes.keys()), 'gene': list(seqs2genes.values())})\n",
    "    cell_barcode_names.to_csv(os.path.join(current_out_path, \"cell_barcode_names.csv\"), header=False)\n",
    "    curr_meta.to_csv(os.path.join(current_out_path, \"meta.csv\"))\n",
    "\n",
    "    expr_dict[current_structure] = cell_by_barcode\n",
    "    # print(curr_meta.head())\n",
    "    sns.distplot(areas)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d4d9c15-4663-44ae-aa54-4eac788ce3cd",
   "metadata": {},
   "source": [
    "## Generate complete matrix for all samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49cb7733-6ccc-41f3-aee7-666958ff90cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct complete matrix\n",
    "sample_dirs = ['ribomap', 'starmap']\n",
    "for current_structure in structure_dict.keys():\n",
    "    print(f\"====Generating: {current_structure}====\")\n",
    "    cell_by_gene_complete = None\n",
    "    meta_complete = None\n",
    "    for i, d in enumerate(sample_dirs):\n",
    "        print(f\"Loading sample: {d}\")\n",
    "        current_expr_path = os.path.join(out_path, d, current_structure)\n",
    "        current_expr = np.loadtxt(os.path.join(current_expr_path, \"cell_barcode_count.csv\"), dtype=int, delimiter=',')\n",
    "        current_meta = pd.read_csv(os.path.join(current_expr_path, \"meta.csv\"))\n",
    "\n",
    "        # add to complete matrix\n",
    "        if cell_by_gene_complete is not None:\n",
    "            cell_by_gene_complete = np.concatenate((cell_by_gene_complete, current_expr))\n",
    "        else:\n",
    "            cell_by_gene_complete = current_expr\n",
    "\n",
    "        if meta_complete is not None:\n",
    "            meta_complete = pd.concat([meta_complete, current_meta])\n",
    "        else:\n",
    "            meta_complete = current_meta\n",
    "\n",
    "    current_output_path = os.path.join(out_path, current_structure)\n",
    "    if not os.path.exists(current_output_path):\n",
    "        os.mkdir(current_output_path)\n",
    "    np.savetxt(os.path.join(current_output_path, \"complete_cell_barcode_count.csv\"), cell_by_gene_complete.astype(np.int), delimiter=',', fmt=\"%d\")\n",
    "    meta_complete = meta_complete.reset_index(drop=True)\n",
    "    meta_complete = meta_complete.rename(columns={\"Unnamed: 0\": \"orig_index\"})\n",
    "    meta_complete.to_csv(os.path.join(current_output_path, \"complete_meta.csv\"))\n",
    "    cell_barcode_names.to_csv(os.path.join(current_output_path, \"cell_barcode_names.csv\"), header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
